### PR TITLE
Do not attempt to ascribe projections out of a ty var

### PR DESCRIPTION
**Pull request by pnkfelix**
If we encounter `_` ascribed to structural pattern like `(a, b)`, just skip relate_types.

Fix #55552

### CODE REVIEW

File: src/librustc/mir/tcx.rs [hunk 1]
```diff
@@ -80,19 +80,21 @@ impl<'a, 'gcx, 'tcx> PlaceTy<'tcx> {
...
-    pub fn projection_ty_core<V, T>(self,
-                                    tcx: TyCtxt<'a, 'gcx, 'tcx>,
-                                    elem: &ProjectionElem<'tcx, V, T>,
-                                    mut handle_field: impl FnMut(&Self, &Field, &T) -> Ty<'tcx>)
-                                    -> PlaceTy<'tcx>
+    pub fn projection_ty_core<V, T, E>(
+        self,
+        tcx: TyCtxt<'a, 'gcx, 'tcx>,
+        elem: &ProjectionElem<'tcx, V, T>,
+        mut handle_field: impl FnMut(&Self, &Field, &T) -> Result<Ty<'tcx>, E>)
```
Comment (nikomatsakis) [inline src/librustc/mir/tcx.rs]: Nit: rustfmt would put the `)` on the next line.

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
@pnkfelix and I were discussing this PR and we observed that it could cause problems if you have repeated type variables, e.g., `type Foo<T> = (T, T)` and an annotation like `Foo<_>`

**Comment by pnkfelix** [comment]
discussed at T-compiler meeting. beta-accepting (with a wee bit of trepidation from @nagisa )

**Comment by pnkfelix** [comment]
```
---- support::resolver::meta_test_multiple_versions_strategy stdout ----
thread 'support::resolver::meta_test_multiple_versions_strategy' panicked at 'In 640 tries we did not see a wide enough distribution of multiple versions of the same library! dis: [387, 79, 51, 20, 16, 15, 2, 0, 3, 12]', tools\cargo\tests\testsuite\support\resolver.rs:484:5
failures:
    support::resolver::meta_test_multiple_versions_strategy
test result: FAILED. 1440 passed; 1 failed; 4 ignored; 0 measured; 0 filtered out
error: test failed, to rerun pass '--test testsuite'
```

... that message ... sounds like a spurious failure to me ...?

File: src/librustc_mir/borrow_check/nll/type_check/mod.rs [hunk 2]
```diff
@@ -991,20 +991,39 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
                 let v1 = ty::Contravariant.xform(v);
 
                 let tcx = self.infcx.tcx;
-                let mut projected_ty = PlaceTy::from_ty(ty);
+                let ty = self.normalize(ty, locations);
+
+                // We need to follow any provided projetions into the type.
```
Comment (bjorn3) [inline src/librustc_mir/borrow_check/nll/type_check/mod.rs]:
```suggestion
                // We need to follow any provided projections into the type.
```

File: src/librustc_traits/type_op.rs [hunk 3]
```diff
@@ -151,17 +151,35 @@ impl AscribeUserTypeCx<'me, 'gcx, 'tcx> {
         debug!("relate_type_and_user_type: ty of def-id is {:?}", ty);
         let ty = self.normalize(ty);
 
-        let mut projected_ty = PlaceTy::from_ty(ty);
+        // We need to follow any provided projetions into the type.
```
Comment (bjorn3) [inline src/librustc_traits/type_op.rs]:
```suggestion
        // We need to follow any provided projections into the type.
```

File: src/test/ui/borrowck/issue-55552-ascribe-wildcard-to-structured-pattern.rs [hunk 4]
```diff
@@ -0,0 +1,31 @@
+// compile-pass
+
+// rust-lang/rust#55552: The strategy pnkfelix landed in PR #55274
+// (for ensuring that NLL respects user-provided lifetime annotations)
+// did not handle the case where the ascribed type has some expliit
```
Comment (bjorn3) [inline src/test/ui/borrowck/issue-55552-ascribe-wildcard-to-structured-pattern.rs]:
```suggestion
// did not handle the case where the ascribed type has some explicit
```

**Review by bjorn3 — COMMENTED** [review]
Noticed some spell errors.

**Pull request merged**